### ISSUE TITLE
Bug report: AirVPN fails to download and install correctly

### ISSUE THREAD

**Issue by Tomk3n** [caskroom/homebrew-cask#28348]
#### AirVPN fails to install

AirVPN isn't downloaded correctly with the command brew cask install airvpn and because of that fails to install. But if downloaded manually the filename is actually the one the cask expects.

<details><summary>Output of `brew cask <command> --verbose`</summary>

==> Downloading https://airvpn.org/eddie/download/?platform=osx&arch=x64&format=installer.pkg&version=2.10
/usr/bin/curl --fail --remote-time --location --user-agent Homebrew/1.1.5 (Macintosh; Intel macOS 10.12.2) curl/7.51.0 https://airvpn.org/eddie/download/?platform=osx&arch=x64&format=installer.pkg&version=2.10 -C 0 -o /Users/tomk3n/Library/Caches/Homebrew/Cask/airvpn--2.10.10.incomplete
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 8573k  100 8573k    0     0  3390k      0  0:00:02  0:00:02 --:--:-- 3391k
==> Verifying checksum for Cask airvpn
==> Running installer for airvpn; your password may be necessary.
==> Package installers may write to any location; options such as --appdir are ignored.
Error: pkg source file not found: '/usr/local/Caskroom/airvpn/2.10/airvpn_osx_x64_installer.pkg'
Error: An error occured during installation of Cask airvpn: pkg source file not found: '/usr/local/Caskroom/airvpn/2.10/airvpn_osx_x64_installer.pkg'
Error: nothing to install
Error: Kernel.exit

</details>

<details><summary>Output of `brew doctor`</summary>

Your system is ready to brew.

</details>

<details><summary>Output of `brew cask doctor`</summary>

==> macOS Release:
10.12.2
==> Hardware Architecture:
intel-64
==> Ruby Version:
2.0.0-p648
==> Ruby Path:
/System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/bin/ruby
==> Homebrew Version:
Homebrew 1.1.5
Homebrew/homebrew-core (git revision 84c7; last commit 2016-12-28)
==> Homebrew Executable Path:
/usr/local/bin/brew
==> Homebrew Cellar Path:
/usr/local/Cellar
==> Homebrew Repository Path:
/usr/local/Homebrew
==> Homebrew Origin:
https://github.com/Homebrew/brew
==> Homebrew-Cask Version:
Homebrew-Cask 1.1.5
caskroom/homebrew-cask (git revision a1eed; last commit 2016-12-28)
==> Homebrew-Cask Install Location:
<NONE>
==> Homebrew-Cask Staging Location:
/usr/local/Caskroom
==> Homebrew-Cask Cached Downloads:
/Users/tomk3n/Library/Caches/Homebrew/Cask
1 files, 8.4M (warning: run "brew cask cleanup")
==> Homebrew-Cask Default Tap Path:
/usr/local/Homebrew/Library/Taps/caskroom/homebrew-cask
==> Homebrew-Cask Alternate Cask Taps:
==> Homebrew-Cask Default Tap Cask Count:
3496
==> Contents of $LOAD_PATH:
/usr/local/Homebrew/Library/Homebrew/cask/lib
/usr/local/Homebrew/Library/Homebrew
/Library/Ruby/Site/2.0.0
/Library/Ruby/Site/2.0.0/x86_64-darwin16
/Library/Ruby/Site/2.0.0/universal-darwin16
/Library/Ruby/Site
/System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/lib/ruby/vendor_ruby/2.0.0
/System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/lib/ruby/vendor_ruby/2.0.0/x86_64-darwin16
/System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/lib/ruby/vendor_ruby/2.0.0/universal-darwin16
/System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/lib/ruby/vendor_ruby
/System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/lib/ruby/2.0.0
/System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/lib/ruby/2.0.0/x86_64-darwin16
/System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/lib/ruby/2.0.0/universal-darwin16
==> Contents of $RUBYLIB Environment Variable:
<NONE>
==> Contents of $RUBYOPT Environment Variable:
<NONE>
==> Contents of $RUBYPATH Environment Variable:
<NONE>
==> Contents of $RBENV_VERSION Environment Variable:
<NONE>
==> Contents of $CHRUBY_VERSION Environment Variable:
<NONE>
==> Contents of $GEM_HOME Environment Variable:
<NONE>
==> Contents of $GEM_PATH Environment Variable:
<NONE>
==> Contents of $BUNDLE_PATH Environment Variable:
<NONE>
==> Contents of $PATH Environment Variable:
PATH="/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/usr/local/Homebrew/Library/Homebrew/cask/cmd:/usr/local/Homebrew/Library/Homebrew/shims/scm"
==> Contents of $SHELL Environment Variable:
SHELL="/bin/bash"
==> Contents of Locale Environment Variables:
LC_CTYPE="UTF-8"

</details>

**Comment by miccal** [comment]
I can reproduce this:
```
-bash-4.4.5- /Users/miccal [> brew cask fetch airvpn
==> Downloading external files for Cask airvpn
==> Downloading https://airvpn.org/eddie/download/?platform=osx&arch=x64&format=installer.pkg&version=2.10
######################################################################## 100.0%
==> Verifying checksum for Cask airvpn
==> Success! Downloaded to -> /Users/miccal/Library/Caches/Homebrew/Cask/airvpn--2.10.10
```
download a plain file, not the required `.pkg`. Entering the url `https://airvpn.org/eddie/download/?platform=osx&arch=x64&format=installer.pkg&version=2.10` does download the `.pkg` correctly.

I do not know what to change in the [Cask file](https://github.com/caskroom/homebrew-cask/blob/master/Casks/airvpn.rb) though?

**Comment by Tomk3n** [comment]
It seems like cask uses .10 as the extension instead of .pkg maybe we could set container type: :pkg to manually tell cask that the downloaded file is a pkg.

**Comment by Numbermaniac** [comment]
The .10 is part of the version number, 2.10, and is being given to the page as a list of arguments. @miccal, is it a completely blank file, or is it just missing the pkg extension?

**Comment by miccal** [comment]
@Numbermaniac it downloads a text file called `airvpn--2.10.10` that when opened is just gibberish.

The [spideroakone Cask](https://github.com/caskroom/homebrew-cask/blob/master/Casks/spideroakone.rb) is similar, and downloads a text file called `spideroakone--latest` which is also gibberish, but `brew cask install spideroakone` works successfully.

**Comment by Numbermaniac** [comment]
@miccal Can you try editing the file extension to add .pkg and see if it opens the installer when double clicked?

**Comment by miccal** [comment]
@Numbermaniac that works, but it does not explain why the Cask is failing when the similar one for spideroakone works as expected.

Also remember that the Cask file expects the package to be called `airvpn_osx_x64_installer.pkg`, just as the Cask file for spideroakone expects the package to be called `SpiderOakONE.pkg`. 

So just appending `.pkg` to the cached file will not fix the core issue.

**Comment by Numbermaniac** [comment]
I know it won't fix the issue, but I wanted to check if it was actually downloading the .pkg at least. So we know it's downloading the correct file, just with a wrong name and extension.

**Comment by Numbermaniac** [comment]
Does homebrew-cask determine the file extension by looking at the URL? Because the URL ends in .10...?

**Comment by reitermarkus** [comment]
Homebrew Cask mostly relies on “magic numbers” to determine the file type. Sadly, `pkg` installers don't have such a “magic number“, so we rely on the extension being `.pkg`, that's why it fails.

**Comment by miccal** [comment]
@reitermarkus so how can we fix this? Why does the spideroakone Cask work, but this one doesn't?

**Comment by reitermarkus** [comment]
`spideroakone` works because it downloads a `.dmg`, not a stand-alone `.pkg`.

**Comment by miccal** [comment]
I see - so would adding `container type: :pkg` work?

**Comment by reitermarkus** [comment]
I'm not sure, if I remember correctly, `installer` also requires a `.pkg` extension, but you could try.

**Comment by miccal** [comment]
@Tomk3n could you try installing AirVPN by running `brew cask install https://raw.githubusercontent.com/caskroom/homebrew-cask/909a469c475eeef046787f0bcaf46cc31cadb863/Casks/airvpn.rb` and report back here of to works?

**Comment by miccal** [comment]
@reitermarkus changing the `url` to `"https://airvpn.org/eddie/download/?platform=osx&arch=x64&version=#{version}&format=installer.pkg"` did not work - it downloaded a file in the cache called `airvpn--2.10.pkg`, but then:
```
==> Downloading https://airvpn.org/eddie/download/?platform=osx&arch=x64&version=2.10&format=installer.pkg
######################################################################## 100.0%
==> Verifying checksum for Cask airvpn
==> Running installer for airvpn; your password may be necessary.
==> Package installers may write to any location; options such as --appdir are ignored.
Error: pkg source file not found: '/usr/local/Caskroom/airvpn/2.10/airvpn_osx_x64_installer.pkg'
Error: An error occured during installation of Cask airvpn: pkg source file not found: '/usr/local/Caskroom/airvpn/2.10/airvpn_osx_x64_installer.pkg'
Error: nothing to install
```

**Comment by reitermarkus** [comment]
`airvpn_osx_x64_installer.pkg` should be `airvpn--#{version}.pkg` now.

**Comment by miccal** [comment]
@reitermarkus thanks, but:
```
==> Downloading https://airvpn.org/eddie/download/?platform=osx&arch=x64&version=2.10&format=installer.pkg
######################################################################## 100.0%
==> Verifying checksum for Cask airvpn
==> Running installer for airvpn; your password may be necessary.
==> Package installers may write to any location; options such as --appdir are ignored.
Error: pkg source file not found: '/usr/local/Caskroom/airvpn/2.10/airvpn--2.10.pkg'
Error: An error occured during installation of Cask airvpn: pkg source file not found: '/usr/local/Caskroom/airvpn/2.10/airvpn--2.10.pkg'
Error: nothing to install
```
Of course the file `airvpn--2.10.pkg` resides in `/Users/$USER/Library/Caches/Homebrew/Cask`, but is not moved to '/usr/local/Caskroom/airvpn/2.10/airvpn--2.10.pkg'.

**Comment by reitermarkus** [comment]
Ah, when it is moving the file, is is extracting the name from the URL again, which results in `/usr/local/Caskroom/airvpn/2.10/download` instead of `/usr/local/Caskroom/airvpn/2.10/airvpn--2.10.pkg`.

**Comment by miccal** [comment]
But when it errors out, there is no `/usr/local/Caskroom/airvpn/` - I assume if it fails, it automatically deletes it?

**Comment by miccal** [comment]
So is it as simple as changing `pkg "airvpn--#{version}.pkg"` to `pkg "download/airvpn--#{version}.pkg"`?

**Comment by reitermarkus** [comment]
It should be as simple as changing it to `pkg 'download'`, which isn't really a good solution.

We can either rename it to something saner inside a `preflight` block or wait on https://github.com/Homebrew/brew/issues/1748.

**Comment by miccal** [comment]
OK, let's just wait for https://github.com/Homebrew/brew/pull/1748.

Thanks for all your help @reitermarkus!
